src/impl_constructors.rs

```diff
@@ -171,6 +171,20 @@ where
         }
         eye
     }
+
+    /// Create a 2D matrix from its diagonal
+    pub fn from_diag(diag: &ArrayBase<S, Ix1>) -> Self
+    where
+        S: DataMut,
+        A: Clone + Zero + One,
+    {
+        let mut eye = Self::zeros((diag.len(), diag.len()));
+        for (idx, val) in diag.iter().enumerate() {
+            eye[[idx, idx]] = val.clone();
```

Comment:
Maybe something like,
```rust
for (a_ii, val) in eye.diag_mut().iter().zip(diag) {
   *a_ii = val.clone();
}
```
would have been better, but I haven't managed to make that work (haven't searched very long). The error was,
```
error[E0594]: cannot assign to `*a_ii` which is behind a `&` reference
   --> src/impl_constructors.rs:183:13
    |
182 |         for (a_ii, val) in eye.diag_mut().iter().zip(diag) {
    |              ---- help: consider changing this to be a mutable reference: `&mut A`
183 |             *a_ii = val.clone()
    |             ^^^^^ `a_ii` is a `&` reference, so the data it refers to cannot be written
```

Reply:
You can fix the error by changing `.iter()` to `.iter_mut()`, but a faster and more concise implementation is to use `.assign()`:

```rust
    pub fn from_diag<S2>(diag: &ArrayBase<S2, Ix1>) -> Self
    where
        A: Clone + Zero,
        S: DataMut,
        S2: Data<Elem = A>,
    {
        let n = diag.len();
        let mut arr = Self::zeros((n, n));
        arr.diag_mut().assign(diag);
        arr
    }
```